[Edge] ICE, DTLS, and sending/receiving audio & video works

ibc opened the pull request:
# ORTC shim for Edge

The `modules/RTC/ortc` folder contains a `RTCPeerConnection` shim for Edge based on ORTC among with other shims (such as `RTCSessionDescription`) and some utililities/helpers to deal with both, SDP and ORTC objects.

## TODO

* WebSocket "shim" for DataChannel not yet done.


## RTCPeerConnection.js

Exports a `RTCPeerConnection` shim.

The interface is based on the [W3C specification of 2015](https://www.w3.org/TR/2015/WD-webrtc-20150210/), which matches (mostly) the current implementation of Chrome.

It also implements Plan-B for multi-stream.


### Limitations

* BUNDLE is assumed (single transport for all the local and remote media streams).
* `rtcp-mux` is assumed (not a real problem nowadays).
* Calling `createOffer()` is not (yet) implemented, so P2P mode is not supported (`RTCBrowserType.isP2PSupported()` returns `false`).
* Simulcast not supported (currently Edge supports `maxFramerate` per encoding but it does not support `maxBitrate`, `resolutionScale` or `framerateScale`, so it's not worth it).
* `RTCDataChannel` not supported (Edge does not implement it).
* `addIceCandidate()` not supported (it's never called anyway, at least in non P2P mode).
  - Edge does not support Trickle-ICE which means that, after reading all the remote ICE candidates from the remote SDP and applying them, an "empty" candidate must be immediately applied (otherwise the `RTCIceTransport` never enters the "completed" state) and, after that, new remote candidates cannot be added.


## RTCSessionDescription.js

Exports a `RTCSessionDescription` shim.

The interface is the same as in the WebRTC specification (although internally it also handles a Object representation of the SDP generated by the [sdp-transform](https://www.npmjs.com/package/sdp-transform) library).


### Limitations

* Cannot use `JSON.stringify()` with a `RTCSessionDescription` instance as argument. Instead, both `type` and `sdp` members must be directly retrieved (this is not a problem within the whole `lib-jitsi-meet` library).


## errors.js

Exports some `Error` based classes needed in WebRTC (such as `InvalidStateError`).


## sdpUtils.js

Some utilities for retrieving information from a SDP.

*NOTE:* One may suggest including these utilities into the already existing `modules/xmpp/SDPUtil.js`, but the fact is that `sdpUtils.js` also produces ORTC objects so it fits better here within the `ortc` folder.


## ortcUtils.js

ORTC related utilities.

ibc commented:
I've pushed a second commit to this PR. Sending audio/video from Edge now works, although the video quality received in Chrome/Firefox is just wrong (not sure if this is a problem in Edge itself or what...).

<img width="1385" alt="captura de pantalla 2017-05-10 a las 20 30 02" src="https://cloud.githubusercontent.com/assets/16191/25914744/9de7068c-35bf-11e7-8631-0b144a0ef06e.png">

gpolitis commented:
Do you always get this? Can you capture a decrypted pcap of the traffic for a 1-1 call that exhibits this behavior?

ibc commented:
> Do you always get this?

Yes.

> Can you capture a decrypted pcap of the traffic for a 1-1 call that exhibits this behavior?

I've not yet implemented P2P mode. It will take some time.

ibc commented:
> I've not yet implemented P2P mode. 

So to clarify: `RTCBrowserType.isP2PSupported()` returns false for Edge.
Said that, audio from Edge (opus) is correctly played in Chrome/Firefox.

gpolitis commented:
sorry, I wasn't talking about P2P mode, there's a way to have the JVB capture decrypted traffic for debugging purposes

    # Packet logging
    org.jitsi.impl.neomedia.transform.DebugTransformEngine.LOG_RTP_PAYLOAD=true
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_ENABLED=true
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_ARBITRARY_ENABLED=true
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_SIP_ENABLED=false
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_JABBER_ENABLED=false
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_RTP_ENABLED=false
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_ICE4j_ENABLED=false
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_FILE_COUNT=2
    net.java.sip.communicator.packetlogging.PACKET_LOGGING_FILE_SIZE=100000000 

This will produce a pcap file with the decrypted traffic.

ibc commented:
Ah ok, thanks. I will check that tomorrow.

saghul commented:
What coded are you supposed to be using? VP8?

ibc commented:
Yes, VP8.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+                    }
+
+                    // Set RTCP feedback.
+                    if (codec.rtcpFeedback) {
```
please reduce this indentation level

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+
+                // If there are no codecs, set this m section as unavailable.
+                if (payloads.length === 0) {
+                    mediaObject.payloads = '7'; // Just put something.
```
`*` ?

ibc replied:
Otherwise payloads in the m line becomes an empty string. I've seen SDP parsers failing on that case. Even Firefox places a custom/random PT when disabling a m section. I'll change it to 9 (it may be any value).

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+            try {
+                rtpSender.send(parameters);
+            } catch (error) {
+                logger.error(`rtpSender.send() failed:${error}`);
```
should the sending flag be reset here?

ibc replied:
If so, it will try to send again for every new SDP O/A (which happens when a new participant enters or the room or leaves it). With same parameters so it will always throw the same error.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+
+        // Check new tracks.
+        for (const pair of newRemoteTrackInfos) {
+            const ssrc = pair[0];
```
use array destructuring

ibc replied:
OK

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+        // Emit MediaStream 'addtrack' for new tracks in already existing
+        // streams.
+        for (const pair of addedRemoteTracks) {
+            const track = pair[0];
```
array destructuring

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+
+            codecPayloadType = codecCapability.preferredPayloadType;
+            codecs.push({
+                name: codecCapability.name,
```
please order these alphabetically

ibc replied:
No please, keys are put in the same order as they appear in the spec. `payloadType` is MUCH more important than `clockRate`, etc.

saghul replied:
Consistency is more important than anything else (to me, that is)

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -36,70 +46,131 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+
+        // Create the RTCDtlsTransport.
+        this._setDtlsTransport(this._iceTransport);
     }
```
Maybe we should check the pcConfig to see if rtcp-mux was attempted to be disabled?

ibc replied:
I don't see any point here. Nobody making WebRTC stuff since 2015 expects non rtcp-mux. Chrome and Firefox are about removing support for it, and rtcp-mux is marked as "implementation risk" in the WG.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+                    `_receiveMedia() new remote stream, id:${streamRemoteId}`);
+
+                // Create a new MediaStream.
+                stream = new MediaStream();
```
add to global?

ibc replied:
Actually linters did not complain. I expect `MediaStream` is already well known stuff.

ibc commented on modules/RTC/ortc/ortcUtils.js:
```diff
@@ -0,0 +1,159 @@
+/* global __filename, RTCRtpReceiver */
```
Yes I do:

```js
const logger = getLogger(__filename);
```

saghul replied:
Not on this file.

ibc replied:
Fair.

ibc commented on modules/RTC/ortc/ortcUtils.js:
```diff
@@ -0,0 +1,159 @@
+                    if (localParamValue === remoteParamValue) {
+                        // Use this RTP parameter.
+                        codec.parameters[localParamName] = localParamValue;
+                        break;
```
No, we just found the mathing param so let's break from the inner `for`. Main `for` stills works.

ibc commented on modules/RTC/ortc/ortcUtils.js:
```diff
@@ -0,0 +1,159 @@
+            }
+            /* eslint-enable max-len */
+
+            for (const remoteFb of remoteCodec.rtcpFeedback) {
```
I don't see how and I don't know why that would be better than current code.

ibc commented on modules/RTC/RTCBrowserType.js:
```diff
@@ -140,7 +140,11 @@ const RTCBrowserType = {
      */
     isP2PSupported() {
-        return !RTCBrowserType.isReactNative();
+        if (RTCBrowserType.isReactNative() || RTCBrowserType.isEdge()) {
```
Will do.

ibc commented on modules/RTC/ortc/sdpUtils.js:
```diff
@@ -0,0 +1,348 @@
+    let media;
+
+    // Ignore disabled media sections.
+    for (const m of sdpObject.media) {
```
Why should I make a generic function just to iterate media sections? Code inside the loop is different in each case.

saghul replied:
I meant to create a function to extract the first non-disabled media stream which has a given property. I count 3 occurrences.

ibc commented on modules/RTC/ortc/sdpUtils.js:
```diff
@@ -0,0 +1,348 @@
+    case 'actpass':
+        role = 'auto';
+        break;
+    }
```
Why? `media.setup` can just be those three values.

saghul replied:
Defensive Programming (R)

ibc commented on modules/RTC/ortc/sdpUtils.js:
```diff
@@ -0,0 +1,348 @@
+        }
+
+        // Get codec parameters.
+        if (Array.isArray(m.fmtp)) {
```
Will do.

ibc replied:
Yes.

ibc commented on modules/RTC/ortc/sdpUtils.js:
```diff
@@ -0,0 +1,348 @@
+                // Don't duplicate.
+                let duplicated = false;
+
+                for (let i = 0, len = headerExtensions.length; i < len; i++) {
```
Yep, doing that elsewhere (where feasible).

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -20,13 +25,18 @@ const RTCIceGatheringState = {
     complete: 'complete'
 };
 
+const CNAME = `jitsi-ortc-cname-${RandomUtil.randomInt(10000, 99999)}`;
```
Ummmm, actually CNAME should be the same for all the PC created by the same physical endpoint.

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -413,39 +514,45 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
     createDataChannel() {
         logger.debug('createDataChannel()');
+
+        // TODO: TBD
```
OK.

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -413,39 +514,45 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
-        // TBD
+        logger.debug('getStats()');
+
+        // TODO: TBD
```
OK

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -528,12 +634,67 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
 
-        // TODO: More stuff to be done.
+        iceTransport.oncandidatepairchange = ev => {
+            logger.debug(
```
No.

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+        this._iceTransport.addRemoteCandidate({});
+
+        // Set desired remote DTLS role (as we receive the offer).
+        // NOTE: Shouldn't be needed since default behaviour (based on ICE
```
OK

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+        // SDP global fields.
+        sdpObject.version = 0;
+        sdpObject.origin = {
+            address: '0.0.0.0',
```
OK

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1052,837 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+            case 'audio':
+            case 'video':
+                mediaObject.protocol = 'RTP/SAVPF';
+                mediaObject.port = 7;
```
Fine.

ibc commented:
Review theoretically satisfied.

saghul commented on modules/RTC/RTCBrowserType.js:
```diff
@@ -226,7 +226,7 @@ const RTCBrowserType = {
         // whether to actually enable simulcast or not).
         // TODO: the logic should be in one single place.
-        return window.chrome !== undefined;
+        return window.chrome !== undefined && !RTCBrowserType.isEdge();
```
Rebase alert! This will probably change soon so you will need to rebase.

ibc replied:
oh, but I already rebased... will check

ibc replied:
I'm afraid. I see `isSimulcastSupported()` and `supportsSimulcast()` (which still checks `window.chrome`) in RTCBrowserType. Help please.

saghul commented on modules/RTC/RTCUtils.js:
```diff
@@ -30,7 +30,7 @@ import VideoType from '../../service/RTC/VideoType';
 
 // Disable Edge until fully implemented.
-const ENABLE_EDGE = false;
+const ENABLE_EDGE = true;
```
Can we just remove this then?

ibc replied:
Right.

saghul commented on modules/RTC/ortc/ortcUtils.js:
```diff
@@ -0,0 +1,153 @@
+
+    // Add RTP header extensions.
+    for (const remoteExtension of filterWithCapabilities.headerExtensions) {
+        for (const localExtension of localFullCapabilities.headerExtensions) {
```
you can reduce a level of indentation here by using Array.some.

ibc replied:
I understand you mean `Array.find`. Done.

saghul replied:
I meant some because there could be more than one.

saghul commented on modules/RTC/ortc/sdpUtils.js:
```diff
@@ -0,0 +1,331 @@
+    return {
+        usernameFragment,
+        password,
+        iceLite
```
alphabetical order

saghul replied:
bump

saghul commented:
I'll test it once I get to my Edge browser. I'd still like to see the alphabetical order thing addressed.

ibc commented:
>  I'd still like to see the alphabetical order thing addressed.

May I talk to my attorney? I just don't agree with that. Elements within ORTC dictionaries are in the same order as in the ORTC spec. Again, putting `clockRate` before `name` is a no sense IMHO.

saghul commented:
> May I talk to my attorney? I just don't agree with that. Elements within ORTC dictionaries are in the same order as in the ORTC spec. Again, putting clockRate before name is a no sense IMHO.

You can talk to anyone, but these are rules and THOU SHALT FOLLOW THEM. 🚓 🚓 🚓 

The reason is simple: the human brain (and computers for that matter) is good at finding words in alphabetical order, finding if an attribute belongs in an unordered list is harder.

saghul commented:
@ibc https://github.com/jitsi/lib-jitsi-meet/pull/478 landed, can you please rebase your PR? The only change that should affect you is the code for simulcast support, which you need to remove, since it's now a lot more explicit.

ibc commented:
> @ibc #478 landed, can you please rebase your PR? The only change that should affect you is the code for simulcast support, which you need to remove, since it's now a lot more explicit.

Done.

ibc commented:
So those are the current issues in the WebRTC shim for Edge. I need to investigate them further. They may be bugs in the shim, bugs in Edge, or issues when talking to the JVB server:

* Sometimes ICE does not connect. Local candidates are gathered and remote candidates are received (as shown by `iceTransport.getRemoteCandidate()`), but local ICE state remains "checking" forever.
* In some cases, remote video takes some seconds to be rendered in Edge.
* VP8 video produced by Edge is terribly rendered in Chrome/Firefox (it may be related to RTP mangling done in the JVB, no idea). See [screenshot](https://github.com/jitsi/lib-jitsi-meet/pull/473#issuecomment-300573271).
* When another browser (Chrome/Firefox) in the conference reloads the web (it correctly dissapears and appears again in Edge) its previous audio/video tracks are correctly stopped and removed, the new audio plays well in Edge, but the new video produces an `InvalidState` error when creating a new `RTCRtpReceiver` for it and calling `rtpReceiver.receive(parameters)`. AFAIS the SSRC values of the new remote video do not match any previous one, but must investigate it further.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -620,28 +840,173 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+        logger.warn('addIceCandidate() not supported');
+
+        // Resolve the promise not to break things.
+        return Promise.resolve();
```
if we are not supporting this, we should reject the promise. This function shouldn't be called anyway in non-P2P mode.

ibc replied:
OK. Indeed that's what I initially did.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1056,831 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+                // Set custom property with the remote id.
+                stream.jitsiRemoteId = streamRemoteId;
+
+                // NOTE: Not needed, jitsiRemoteId does work.
```
please remove the comment then

ibc replied:
Will do.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1056,831 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+                // Set custom property with the remote id.
+                track.jitsiRemoteId = trackRemoteId;
+
+                // NOTE: Not needed, jitsiRemoteId does work.
```
ditto

ibc replied:
Will do.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -703,6 +1900,7 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
 
     /**
      * May emit 'negotiationneeded' event.
+     * @private
```
order these functions alphabetically please

ibc replied:
No please, don't ask me to do that. This is about reading code in a logical way. Would you place an `addItem()` method **before** the `constructor()`?

saghul replied:
Constructor goes first, then everything must go alphabetically sorted.

saghul commented on modules/RTC/ortc/RTCSessionDescription.js:
```diff
@@ -6,20 +6,20 @@ import sdpTransform from 'sdp-transform';
 export default class RTCSessionDescription {
```
How about adding some `toJSON` method to this class so it behaves like the one in Chrome / Firefox?

ibc replied:
AFAIR we already discussed about that and concluded that, since it's not an standard method at all and libjitsi does not use it anywhere, there is no value in adding it.

saghul replied:
I think we might rely on it for logging. At any rate, it doesn't hurt.

saghul commented on modules/RTC/ortc/RTCSessionDescription.js:
```diff
@@ -54,23 +54,38 @@ export default class RTCSessionDescription {
-     * @return {string}
+     * @return {String}
      */
     get sdp() {
```
alphabetical order

ibc commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -36,70 +46,131 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+     * Current ICE+DTLS connection state.
+     * @return {RTCPeerConnectionState}
+     */
+    get connectionState() {
```
No please, those getters have a logical order.

saghul replied:
For the Nth time, alphabetical sort. By getter name. Except the constructor, which always goes first. It's not that hard.

ibc commented:
OK, I've put class getters/setters in alphabetical order, removed commented code, and added the `description.toJSON()` method as requested.

Question: do you mean that I must also put class methods (all but the constructor) in alphabetical order?

ibc commented:
I've also rebased with master.

ibc commented:
> Also, can you update the TODO document with the current state? (some time has passed and it's hard to keep track).

I've committed a very very minor TODO update. Every in `modules/RTC/ortc/README.md` is up to date.

saghul commented on modules/RTC/ortc/README.md:
```diff
@@ -0,0 +1,69 @@
+
+There are currently some issues that prevent this WebRTC implementation to properly work:
+
+* Sometimes ICE does not connect. Local candidates are gathered and remote candidates are received (as shown by `iceTransport.getRemoteCandidate()`), but local ICE state remains "checking" forever.
```
This will need fixing before merging.

ibc replied:
Sure, I will work today on this (if I get Wireshark work on Windows...).

saghul commented on modules/RTC/ortc/README.md:
```diff
@@ -0,0 +1,69 @@
+* Sometimes ICE does not connect. Local candidates are gathered and remote candidates are received (as shown by `iceTransport.getRemoteCandidate()`), but local ICE state remains "checking" forever.
+  - *NOTE:* It may be due a ICE role conflict.
+* In some cases, remote video takes some seconds to be rendered in Edge.
+* VP8 video produced by Edge is terribly rendered in Chrome/Firefox (it may be related to RTP mangling done in the JVB, no idea). See [screenshot](https://github.com/jitsi/lib-jitsi-meet/pull/473#issuecomment-300573271).
```
Ditto.

saghul commented on modules/RTC/ortc/README.md:
```diff
@@ -0,0 +1,69 @@
+
+#### Limitations
+
+* Cannot use `JSON.stringify()` with a `RTCSessionDescription` instance as argument. Instead, both `type` and `sdp` members must be directly retrieved (this is not a problem within the whole `lib-jitsi-meet` library). Also, a `toJSON()` method has been added (not standard, but implemented by Chrome and Firefox).
```
The limitation no longer applies then, lets remove it.

ibc replied:
I think it's worth noting that one cannot do `JSON.stringify(description)` and expect a pure JSON string having just `type` and `sdp`. If you don't care about that confirm it to me and I'l remove the limitation from the README.

saghul commented on modules/RTC/ortc/README.md:
```diff
@@ -0,0 +1,69 @@
+*NOTE:* One may suggest including these utilities into the already existing `modules/xmpp/SDPUtil.js`, but the fact is that `sdpUtils.js` also produces ORTC objects so it fits better here within the `ortc` folder.
+
+
+### ortcUtils.js
```
I wonder if we shouldn't merge these 2 utils files into a simple utils.js ...

ibc replied:
Makes sense. Done.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -1,9 +1,14 @@
+import * as sdpUtils from './sdpUtils';
+import * as ortcUtils from './ortcUtils';
 import { InvalidStateError } from './errors';
+import RandomUtil from '../../util/RandomUtil';
```
add an empty line before this last import

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -36,70 +46,123 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+
+        // RTCDtlsTransport.
+        // @type {RTCDtlsTransport}
+        this._dtlsTransport = null;
```
alpha sort

ibc replied:
Done

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -413,39 +514,49 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+
+        // TODO: TBD
+
+        throw new Error('getStats() not yet implemented');
```
shouldn't this return a rejected promise?

ibc replied:
It depends on the app usage. Old WebRTC API throws when something is wrong, or sometimes invokes the given errback. I'll complete it with both a callback/errback and promise capable signature, as all the other methods.

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -620,28 +840,171 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+     * @return {Promise}
+     * @private
+     */
+    _addIceCandidate(candidate) { // eslint-disable-line no-unused-vars
```
Y U NO ALPHA SORT

saghul commented on modules/RTC/ortc/RTCPeerConnection.js:
```diff
@@ -672,18 +1054,817 @@ export default class ortcRTCPeerConnection extends yaeti.EventTarget {
+
+                    // Set a random provisional SSRC if not set.
+                    if (!info.ssrc) {
+                        info.ssrc = RandomUtil.randomInt(1000000, 9999999);
```
Please use `SDPUtil.generateSsrc()` here.

ibc replied:
ok

ibc commented:
Everything is alphabetically ordered right now.

ibc commented:
Hey, good news! The wrong video from Edge is fixed by setting:

`org.jitsi.videobridge.ENABLE_VP8_PICID_REWRITING=false`

in `/etc/jitsi/videobridge/sip-communicator.properties`.

<img width="1349" alt="captura de pantalla 2017-06-05 a las 17 33 45" src="https://cloud.githubusercontent.com/assets/16191/26790916/c077409e-4a15-11e7-9a3f-d33b49d3838f.png">

saghul commented:
Jenkins, test this please.

saghul commented:
It's with great pleasure that I hit the merge button at last. This was a long time in the making and no easy feat. Thanks a lot for your work here Iñaki! ❤️

ibc commented:
<3!
